### ISSUE TITLE
Stop enrollment for users on specific versions of the core add-on

### ISSUE THREAD

**Issue by Dexterp37** [mozilla-rally/rally-core-addon#132]
When we start using a core add-on to enroll study add-ons, there will likely be cases where we want to stop enrollment for users on some version of the core add-on because it has a bug and must be updated before we can proceed (either legal or technical).

For instance if we need to update our program consent dialog before we can enable certain partner studies. other than that just garden-variety technical bugs.

A technical bug would be like, we're going to accidentally through away data if we let users enroll version z study if the user has core add-on x installed so we need to update to core add-on y before unlocking it, or something like that.

This can be done either in the partner support library or remote settings (depending on the desired latency of days vs minutes).

**Comment by marniepw** [comment]
Have kill switch ability via RS. Does this work with core add-on?

But if need to update study, operational processes here.

Need proposal, can be done here in ticket, but need to figure out ASAP.

**Comment by rhelmer** [comment]
I've [started a proposal](https://docs.google.com/document/d/1hciaUOnMBRDrVyUaLf0Pjor8Zhtx0_BIoSL5dkMcgMQ/edit#) that covers this use case.

**Comment by rhelmer** [comment]
Now that PR #311 has landed we should be able to implement this.

Per https://github.com/mozilla-rally/rally-core-addon/issues/132#issue-736305535 we need to be able to handle pausing and resuming studies for specific version(s) of the core add-on.

I'm thinking of handling this by adding a new attribute to the study metadata, something like:

```js
studyPaused: {
  "a.b.c": true, 
  "x.y.z": false,
}
```

Which would cause the study to be paused on version "a.b.c" but not version "x.y.z" of the core add-on. We could also allow `"*"` which would pause is for all versions of the core add-on.

Technically I am thinking that when a study is "paused" this means:

- it is hidden from the UI (no further enrollments allowed)
- any already-installed study add-ons are disabled

Assuming this all makes sense, the one bit I am unsure about is how disabling study add-ons should be implemented. It feels like the choices here to either:

1. disable the study extension using the browser API (unfortunately the [WebExtension API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/management/setEnabled) only works for themes not other extensions)
2. pass a message and have `rally.js` handle it in the study itself

I think given the lack of a standard API for (1), (2) is probably the way we should go. I think that means exactly what "disabled" means is ambiguous - presumably the study add-on needs to continue to listen for messages but doesn't need to do much else.

@Dexterp37 any thoughts?

**Comment by rhelmer** [comment]
> Technically I am thinking that when a study is "paused" this means:
> 
>     * it is hidden from the UI (no further enrollments allowed)

Hm there's a nuance here too I think - if the user has not joined a study then just hiding paused studies from the UI makes sense, but if the study is installed then we don't want to take away the ability to manage it and unenroll even if it is currently paused.

**Comment by Dexterp37** [comment]
> I'm thinking of handling this by adding a new attribute to the study metadata, something like:
> 
> ```js
> studyPaused: {
>   "a.b.c": true, 
>   "x.y.z": false,
> }
> ```
> 
> This would cause the study to be paused on version "a.b.c" but not version "x.y.z" of the core add-on.

You said `core add-on`, but did you mean study?

> We could also allow `"*"` which would pause is for all versions of the core add-on.

Does this mean we should explicitly write down all the versions (i.e. `"x.y.z": false`)?

> Technically I am thinking that when a study is "paused" this means:
> 
>     * it is hidden from the UI (no further enrollments allowed)

Yup, as you mentioned in [this comment](https://github.com/mozilla-rally/rally-core-addon/issues/132#issuecomment-773042964) we should only hide if not installed.

>     * any already-installed study add-ons are disabled

We should carefully consider the following cases, I think:

- What happens if the paused study is installed from outside of the Core Addon? (Without passing through the core addon UI)
- What happens if we say a study should be paused and we fail to do that?
- What does "paused" mean?
 
> Assuming this all makes sense, the one bit I am unsure about is how disabling study add-ons should be implemented. It feels like the choices here to either:
> 
>     1. disable the study extension using the browser API (unfortunately the [WebExtension API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/management/setEnabled) only works for themes not other extensions)
> 
>     2. pass a message and have `rally.js` handle it in the study itself
> 
> 
> I think given the lack of a standard API for (1), (2) is probably the way we should go. I think that means exactly what "disabled" means is ambiguous - presumably the study add-on needs to continue to listen for messages but doesn't need to do much else.

(2) seems like the best fit here. But we should figure out what does "disable" really means for the addon.

**Comment by rhelmer** [comment]
> > I'm thinking of handling this by adding a new attribute to the study metadata, something like:
> > ```js
> > studyPaused: {
> >   "a.b.c": true, 
> >   "x.y.z": false,
> > }
> > ```
> > 
> > 
> > This would cause the study to be paused on version "a.b.c" but not version "x.y.z" of the core add-on.
> 
> You said `core add-on`, but did you mean study?

From the first comment on this issue:

> A technical bug would be like, we're going to accidentally through away data if we let users enroll version z study if the user has core add-on x installed so we need to update to core add-on y before unlocking it, or something like that.

So I am using that at the use case for having the per-version pausing.

> 
> > We could also allow `"*"` which would pause is for all versions of the core add-on.
> 
> Does this mean we should explicitly write down all the versions (i.e. `"x.y.z": false`)?

I'd use the `"*"` as an implicit alternative to needing to do that. We could be explicit instead.

> > Technically I am thinking that when a study is "paused" this means:
> > ```
> > * it is hidden from the UI (no further enrollments allowed)
> > ```
> 
> Yup, as you mentioned in [this comment](https://github.com/mozilla-rally/rally-core-addon/issues/132#issuecomment-773042964) we should only hide if not installed.
> 
> > ```
> > * any already-installed study add-ons are disabled
> > ```
> 
> We should carefully consider the following cases, I think:
> 
>     * What happens if the paused study is installed from outside of the Core Addon? (Without passing through the core addon UI)
> 
>     * What happens if we say a study should be paused and we fail to do that?
> 
>     * What does "paused" mean?
> 
> 
> > Assuming this all makes sense, the one bit I am unsure about is how disabling study add-ons should be implemented. It feels like the choices here to either:
> > ```
> > 1. disable the study extension using the browser API (unfortunately the [WebExtension API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/management/setEnabled) only works for themes not other extensions)
> > 
> > 2. pass a message and have `rally.js` handle it in the study itself
> > ```
> > 
> > 
> > I think given the lack of a standard API for (1), (2) is probably the way we should go. I think that means exactly what "disabled" means is ambiguous - presumably the study add-on needs to continue to listen for messages but doesn't need to do much else.
> 
> (2) seems like the best fit here. But we should figure out what does "disable" really means for the addon.

+1, I have some thoughts and a patch I am working on so we can discuss the little details more concretely.

**Comment by Dexterp37** [comment]
> 
> 
> > > I'm thinking of handling this by adding a new attribute to the study metadata, something like:
> > > ```js
> > > studyPaused: {
> > >   "a.b.c": true, 
> > >   "x.y.z": false,
> > > }
> > > ```
> > > 
> > > 
> > > This would cause the study to be paused on version "a.b.c" but not version "x.y.z" of the core add-on.
> > 
> > 
> > You said `core add-on`, but did you mean study?
> 
> From the first comment on this issue:
> 
> > A technical bug would be like, we're going to accidentally through away data if we let users enroll version z study if the user has core add-on x installed so we need to update to core add-on y before unlocking it, or something like that.
> 
> So I am using that at the use case for having the per-version pausing.

Ok, thinking of this again, the above use-case might also be handled from within `rally.js`. We actually need to, because version mismatches might happen regardless of what the remote settings say.

So what if we descope/refocus this issue to exclusively pause _studies_?

> > > We could also allow `"*"` which would pause is for all versions of the core add-on.
> > 
> > 
> > Does this mean we should explicitly write down all the versions (i.e. `"x.y.z": false`)?
> 
> I'd use the `"*"` as an implicit alternative to needing to do that. We could be explicit instead.

What I mean is: what happens for version that are not listed there?

**Comment by rhelmer** [comment]
> > > > I'm thinking of handling this by adding a new attribute to the study metadata, something like:
> > > > ```js
> > > > studyPaused: {
> > > >   "a.b.c": true, 
> > > >   "x.y.z": false,
> > > > }
> > > > ```
> > > > 
> > > > 
> > > > This would cause the study to be paused on version "a.b.c" but not version "x.y.z" of the core add-on.
> > > 
> > > 
> > > You said `core add-on`, but did you mean study?
> > 
> > 
> > From the first comment on this issue:
> > > A technical bug would be like, we're going to accidentally through away data if we let users enroll version z study if the user has core add-on x installed so we need to update to core add-on y before unlocking it, or something like that.
> > 
> > 
> > So I am using that at the use case for having the per-version pausing.
> 
> Ok, thinking of this again, the above use-case might also be handled from within `rally.js`. We actually need to, because version mismatches might happen regardless of what the remote settings say.
> 
> So what if we descope/refocus this issue to exclusively pause _studies_?

I've only been thinking about pausing studies so far :) The version scheme would just be if we want to communicate to the core add-on that a study is not to be resumed unless the core add-on is a sufficient version. However we could handle this through `rally.js` as you mention.

I think building a complex targeting system is probably not worth it at this stage, it's probably sufficient to communicate to the core add-on which studies are available, and of those which are currently in a "paused" state.

The idea of pausing the core add-on *itself* I haven't really given much thought to, other than the per-version targeting to stop individual studies.

> 
> > > > We could also allow `"*"` which would pause is for all versions of the core add-on.
> > > 
> > > 
> > > Does this mean we should explicitly write down all the versions (i.e. `"x.y.z": false`)?
> > 
> > 
> > I'd use the `"*"` as an implicit alternative to needing to do that. We could be explicit instead.
> 
> What I mean is: what happens for version that are not listed there?

Oh, I see. We could either let them fall back to the default (`paused == false`) or something like:

```js
studyPaused: {
  "x.y.z": true,
  "*": false,
}
```

So in that case the study would only be paused if the current core add-on version is `x.y.z` otherwise it will be enabled. However as discussed above maybe this is too complex and not worth worrying about, vs. something simple like:

```js
studyPaused: true
```

Which simply pauses the study regardless of the version of the core add-on installed. I would guess that this is going to be by far the most common use case of a "kill switch" sort of feature.

**Comment by Dexterp37** [comment]
> ```js
> studyPaused: true
> ```
> 
> Which simply pauses the study regardless of the version of the core add-on installed. I would guess that this is going to be by far the most common use case of a "kill switch" sort of feature.

I would go for this simplest case for now, for the following reasons:

1. Time constraints for Alpha.
2. Targeting by version would require us to have a stricter control/guidelines for versioning, which we don't have yet.

@knowtheory , thoughts?

**Comment by dveditz** [comment]
> pass a message and have rally.js handle it in the study itself

While the study probably wants to know its been paused so it can stop doing work, you can't fully trust rally.js in the study as an enforcement. The core addon needs to also ignore messages from the study. Or at least most messages (there might be "management" type things to communicate in the future?) and definitely stop passing on telemetry pings from paused studies.

**Comment by rhelmer** [comment]
#386 adds a remote-controllable way to pause and resume studies, which addresses this issue.

The only detail that we have not directly tackled is the summary, "Stop enrollment for users on *specific versions* of the core add-on" - the above PR instead allows a way to pause and resume studies regardless of the version of the core add-on.

I think if we encounter a severe bug in the core add-on *itself*, I think a better approach for now would be to serve an add-on update. We can assume in general that users are getting the add-on updates in a timely manner, and if it is misbehaving then it might not listen to our remote-settings changes and acting on them inappropriately, anyway.

**Comment by rhelmer** [comment]
> > pass a message and have rally.js handle it in the study itself
> 
> While the study probably wants to know its been paused so it can stop doing work, you can't fully trust rally.js in the study as an enforcement. The core addon needs to also ignore messages from the study. Or at least most messages (there might be "management" type things to communicate in the future?) and definitely stop passing on telemetry pings from paused studies.

+1, we should be able to tell which messages are coming from which study add-on via `[runtime.MessageSender](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/MessageSender)`.

I'll address this in #386 - we can drop messages from study add-ons that are trying to send telemetry pings while in a pause state. We don't want to block all messages as you suggest, since we need to be able to detect that it is installed and user has consented to the study and so on.

**Comment by Dexterp37** [comment]
This was fixed by #386
